Feature/hamiltonian generation

Stellogic opened the pull request:
This pull request implements **Task 1** of the follow-up plan, introducing utility functions to generate common spin Hamiltonians from `Lattice` objects.

### Key Contributions:

- **`generate_heisenberg_hamiltonian`**: Creates a sparse Heisenberg Hamiltonian based on the nearest-neighbor pairs provided by the lattice.
- **`generate_rydberg_hamiltonian`**: Creates a sparse Rydberg Hamiltonian, utilizing the lattice's distance matrix to calculate long-range interactions.
- **Comprehensive Tests**: Added a new test file (`tests/templates/test_hamiltonians.py`) with unit tests that verify:
    - Scientific correctness against manually calculated small systems.
    - Robustness against edge cases (empty lattice, single site, zero distance).
    - Fundamental physical properties (e.g., Hermiticity).

refraction-ray commented on tensorcircuit/templates/hamiltonians.py:
```diff
@@ -0,0 +1,139 @@
+from .lattice import AbstractLattice
+
+
+def generate_heisenberg_hamiltonian(
```
`def heisenberg_hamiltonian`

refraction-ray commented on tensorcircuit/templates/hamiltonians.py:
```diff
@@ -0,0 +1,139 @@
+
+
+def generate_heisenberg_hamiltonian(
+    lattice: AbstractLattice, j_coupling: float = 1.0
```
j_coupling is better be a vector whose dimension can be 1, 3, representing for 1: $J_x=J_y=Jz$; 3 different $J_{x, y, z}$

refraction-ray commented on tensorcircuit/templates/hamiltonians.py:
```diff
@@ -0,0 +1,139 @@
+            ls.append(zz_string)
+            weights.append(coefficient)
+
+            z_coefficients[i] -= coefficient
```
what is this for?

refraction-ray replied:
ah, i see, my bad

refraction-ray commented on tests/test_hamiltonians.py:
```diff
@@ -0,0 +1,132 @@
+        """
+        Test properties of a larger lattice (2x2 square) without full matrix comparison.
+        """
+        lattice = SquareLattice(size=(2, 2), pbc=True)  # 4 sites, 8 bonds with PBC
```
2*2 with PBC is subtle, as you can see, there will be two bonds between the same two lattice points

refraction-ray commented on tensorcircuit/templates/hamiltonians.py:
```diff
@@ -0,0 +1,139 @@
+
+def generate_heisenberg_hamiltonian(
+    lattice: AbstractLattice, j_coupling: float = 1.0
+) -> coo_matrix:
```
for simplicity, return type coo_matrix=Any is acceptable

refraction-ray commented on tensorcircuit/templates/hamiltonians.py:
```diff
@@ -84,12 +92,12 @@ def generate_rydberg_hamiltonian(
     num_sites = lattice.num_sites
     if num_sites == 0:
-        return coo_matrix((0, 0))
+        return _create_empty_sparse_matrix(shape=(1, 1))
```
similarly, just raise ValueError for illegal number of sites

Stellogic commented:
Hi @refraction-ray,
Thank you for the detailed and helpful review! I have updated the code according to your feedback.
Here is a summary of the changes:
Internal API Usage: I've updated the code to use backend from .cons and relative imports for internal modules, instead of calling the top-level tensorcircuit package.
Error Handling: I've added a ValueError check for num_sites = 0 in both Hamiltonian functions, as you suggested.
Rydberg Hamiltonian Sign: Regarding your question on the Z-term sign (why -v_ij/4?), I've double-checked the derivation and I thought the implementation is correct. The single-body Z coefficient comes from two sources:
The detuning term -delta * n_i gives a + (delta / 2) * Z_i contribution.
The interaction term V_ij * n_i * n_j expands to (V_ij / 4) * (1 - Z_i - Z_j + Z_i*Z_j), which contributes - (V_ij / 4) * Z_i to the single-body term.
I believe this addresses all your feedback. The PR should be ready for another look. Thanks again

refraction-ray commented on tensorcircuit/templates/hamiltonians.py:
```diff
@@ -0,0 +1,154 @@
+import typing
+from typing import Any, List, Tuple, Union
+import numpy as np
+from tensorcircuit.cons import dtypestr, backend
```
should also be `from ..cons` ?

refraction-ray commented on tensorcircuit/templates/hamiltonians.py:
```diff
@@ -0,0 +1,154 @@
+import typing
```
should merge with the following line? import all types you will use

Stellogic commented:
Thanks for the review, @refraction-ray! I've updated the PR according to your suggestions. It should be ready for another look now.

refraction-ray commented:
great! LGTM. will merge if CI finished successfully.
